### PR TITLE
Review HtmlHelper's conversion process #4534

### LINKED ISSUE #4534: Review HtmlHelper's conversion process
It all started with the intention to fix the inconsistent comment handling in the said process. As I examined `HtmlHelper`, I found out that there are so many more inconsistent/redundant/wrong processes inside, so I might as well fix them all for good here.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #4534 
Fixing this now because I made a promise to do it by this milestone. I hope to still fulfill it (or at most overshoot by 1-2 days).

### CODE REVIEW

**Comment by wkurniawan07** [comment]
Will make a gist to explain all the fixes that I've done. Wait for a while.

**Comment by kanghj** [comment]
I don't think I will have enough time to look through this properly any time soon. @thyageshm or @bamboo3250, do you want to review this?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 1]
```diff
@@ -112,146 +89,147 @@ private static Node getNodeFromString(String string) throws SAXException, IOExce
...
-        for (int i = 0; i < actualAttributeList.getLength(); i++){
-            nodesList.add(actualAttributeList.item(i));
+    private static String getNodeOpeningTag(Node currentNode) {
+        StringBuilder openingTag = new StringBuilder();
+        // add the start of opening tag
+        openingTag.append("<" + currentNode.getNodeName().toLowerCase());
+        
+        // add the attributes of the tag (getAttributes() returns the attributes sorted alphabetically)
+        NamedNodeMap attributes = currentNode.getAttributes();
+        if (attributes != null) { // TODO check if this null check is still necessary
```
Comment (kanghj) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I think this check isn't necessary.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 2]
```diff
@@ -112,146 +89,147 @@ private static Node getNodeFromString(String string) throws SAXException, IOExce
...
+            for (int i = 0; i < attributes.getLength(); i++) {
+                Node attribute = attributes.item(i);
+                openingTag.append(" " + attribute.getNodeName().toLowerCase() + "="
+                                      + "\"" + attribute.getNodeValue().replace("\"", "&quot;") + "\"");
```
Comment (kanghj) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I don't really understand this. If we need to replace " with &quot; here, shouldn't the replacements for other characters in sanitizeForHtml be done here too?

Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: This is not for sanitization, but for `DOMParser` to be able to parse properly. Parsing something like `value="abc"def"` could be disastrous, so `value="abc&quot;def"`. Replacing with `'` (as before) doesn't make sense.

Comment (kanghj) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Ok I understand.

**Comment by wkurniawan07** [comment]
I took some time to study [this](https://docs.oracle.com/javase/7/docs/api/org/w3c/dom/Node.html), and now managed to implement a much better solution. Will update the gist.

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 1 above)
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I managed to fix the whole thing to make it unnecessary after all. Thanks for the encouragement!

**Comment by damithc** [comment]
I guess this gist explains to current developers how the solution was arrived at. Do we have a separate documentation on how the current solution works, for future developers?.

**Comment by wkurniawan07** [comment]
In the class itself I've put some comments (both as method signatures and inside the method) to explain the processes. If needed (and I think it will be good) I can make an in-depth technical report on how we handle HTML comparisons (in particular `HtmlHelper` conversion process and God mode). By now I know the process like the back of my hand already.

**Pull request merged**